src/core/scalar.rs

```diff
@@ -0,0 +1,312 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+
+use lazy_static::lazy_static;
+use schemars::schema::{InstanceType, Schema, SchemaObject};
+
+use crate::core::json::JsonLike;
+
+#[derive(schemars::JsonSchema, Debug, Clone, strum_macros::Display, strum_macros::EnumIter)]
+pub enum ScalarType {
+    Empty,
+    Email,
+    PhoneNumber,
+    Date,
+    Url,
+    JSON,
+    Int8,
+    Int16,
+    Int32,
+    Int64,
+    Int128,
+    UInt8,
+    UInt16,
+    UInt32,
+    UInt64,
+    UInt128,
+    Bytes,
+}
+
+fn eval_str<'a, Value: JsonLike<'a> + 'a, F: Fn(&str) -> bool>(val: &'a Value, fxn: F) -> bool {
+    val.as_str().map_or(false, fxn)
+}
+
+fn eval_signed<
+    'a,
+    Num,
+    Value: JsonLike<'a> + 'a,
+    F: Fn(i64) -> Result<Num, std::num::TryFromIntError>,
+>(
+    val: &'a Value,
+    fxn: F,
+) -> bool {
+    val.as_i64().map_or(false, |n| fxn(n).is_ok())
+}
+
+fn eval_unsigned<
+    'a,
+    Num,
+    Value: JsonLike<'a> + 'a,
+    F: Fn(u64) -> Result<Num, std::num::TryFromIntError>,
+>(
+    val: &'a Value,
+    fxn: F,
+) -> bool {
+    val.as_u64().map_or(false, |n| fxn(n).is_ok())
+}
+
+impl ScalarType {
+    pub fn validate<'a, Value: JsonLike<'a> + 'a>(&self, value: &'a Value) -> bool {
+        match self {
+            ScalarType::JSON => true,
+            ScalarType::Empty => true,
+            ScalarType::Email => eval_str(value, |s| {
+                async_graphql::validators::email(&s.to_string()).is_ok()
+            }),
+            ScalarType::PhoneNumber => eval_str(value, |s| phonenumber::parse(None, s).is_ok()),
+            ScalarType::Date => {
+                eval_str(value, |s| chrono::DateTime::parse_from_rfc3339(s).is_ok())
+            }
+            ScalarType::Url => eval_str(value, |s| url::Url::parse(s).is_ok()),
+            ScalarType::Bytes => value.as_str().is_some(),
+
+            ScalarType::Int64 => eval_str(value, |s| s.parse::<i64>().is_ok()),
+            ScalarType::UInt64 => eval_str(value, |s| s.parse::<u64>().is_ok()),
+            ScalarType::Int128 => eval_str(value, |s| s.parse::<i128>().is_ok()),
+            ScalarType::UInt128 => eval_str(value, |s| s.parse::<u128>().is_ok()),
+
+            ScalarType::Int8 => eval_signed(value, i8::try_from),
+            ScalarType::Int16 => eval_signed(value, i16::try_from),
+            ScalarType::Int32 => eval_signed(value, i32::try_from),
+
+            ScalarType::UInt8 => eval_unsigned(value, u8::try_from),
+            ScalarType::UInt16 => eval_unsigned(value, u16::try_from),
+            ScalarType::UInt32 => eval_unsigned(value, u32::try_from),
+        }
+    }
+    pub fn get_scalar(name: &str) -> ScalarType {
+        CUSTOM_SCALARS.get(name).cloned().unwrap_or(Self::Empty)
```

Comment:
Should return an option, let the caller decide what to do if nothing is found. Add docs.

Reply:
if scalar is not found then validate function should always return `true` by default